### PR TITLE
NFDIV-743

### PR DESCRIPTION
**Pull request by tomxelliott**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/NFDIV-743

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/divorce/citizen/event/CitizenLinkApplicationTest.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @AfterAll
+    static void tearDown() {
+        IdamWireMock.stopAndReset();
+        CaseDataWireMock.stopAndReset();
+    }
+
+    @Test
+    public void givenValidAccessCodeWhenCallbackIsInvokedThenAccessCodeIsRemovedAndSolicitorRolesAreSet() throws Exception {
```
Comment (nitinprabhu) [inline src/integrationTest/java/uk/gov/hmcts/divorce/citizen/event/CitizenLinkApplicationTest.java:93]: could we cover unhappy path also ?

File: src/main/java/uk/gov/hmcts/divorce/common/model/access/DefaultAccess.java [hunk 2]
```diff
@@ -11,13 +11,15 @@
...
 import static uk.gov.hmcts.divorce.common.model.UserRole.CASEWORKER_DIVORCE_SUPERUSER;
+import static uk.gov.hmcts.divorce.common.model.UserRole.CASEWORKER_DIVORCE_SYSTEMUPDATE;
 import static uk.gov.hmcts.divorce.common.model.UserRole.CITIZEN;
 
 public class DefaultAccess implements HasAccessControl {
     @Override
     public SetMultimap<HasRole, Permission> getGrants() {
         SetMultimap<HasRole, Permission> grants = HashMultimap.create();
         grants.putAll(CITIZEN, Permissions.CREATE_READ_UPDATE);
+        grants.putAll(CASEWORKER_DIVORCE_SYSTEMUPDATE, Permissions.CREATE_READ_UPDATE);
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/divorce/common/model/access/DefaultAccess.java:22]: why do we need this role to be added ?

Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/divorce/common/model/access/DefaultAccess.java:22]: The system caseworker user who is invoking the linking for joint application does not have the CRU permissions for all Casefields so CCD Data Store was rejecting the request. We need the system case worker to have the same access to the citizen case fields as a citizen user.

File: src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenLinkApplication.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+import static org.springframework.http.HttpHeaders.AUTHORIZATION;
+import static uk.gov.hmcts.divorce.common.model.State.AwaitingApplicant2Response;
+import static uk.gov.hmcts.divorce.common.model.UserRole.CASEWORKER_DIVORCE_SYSTEMUPDATE;
+import static uk.gov.hmcts.divorce.common.model.UserRole.CITIZEN;
+import static uk.gov.hmcts.divorce.common.model.access.Permissions.CREATE_READ_UPDATE;
+
+@Slf4j
+@Component
+public class CitizenLinkApplication implements CCDConfig<CaseData, State, UserRole> {
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenLinkApplication.java:25]: Maybe it should be `CitizenLinkApplicant2`?

File: src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenLinkApplication.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    public static final String CITIZEN_LINK_APPLICANT_2 = "citizen-link-applicant2";
+
+    @Override
+    public void configure(final ConfigBuilder<CaseData, State, UserRole> configBuilder) {
+        configBuilder
+            .event(CITIZEN_LINK_APPLICANT_2)
+            .forState(AwaitingApplicant2Response)
+            .name("Link Applicant 2 to case")
+            .description("Link Applicant 2 to case to enable completion of joint application")
+            .grant(CREATE_READ_UPDATE, CITIZEN, CASEWORKER_DIVORCE_SYSTEMUPDATE)
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenLinkApplication.java:42]: A citizen shouldn't be able to trigger this - it has to be the system.

**Review by linusnorton — APPROVED** [review]
Assuming the CITIZEN permission for the link-applicant2 event is taken away it's all good

File: src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenLinkApplication.java (hunk 4 above)
Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenLinkApplication.java:42]: Will update this

**Pull request merged**